File: src/summation_primes.c
/* Finding the summation of primes < 2mil
 *
 * Answer: 142913828922
 *
 * Project Euler: 10 */

#include <stdio.h>
#include <stdint.h>
#include <inttypes.h>
#include <math.h>

#define MAX 2000000

int main(int argc, char** argv);
int is_prime(int32_t s);


int main(int argc, char** argv)
{
        (void) argc;
        (void) argv;

        int64_t sum = 2;

        for(int32_t i = 3; i < MAX; i += 2){

                sum += (is_prime(i)) ? i : 0;
        }

        printf("%"PRId64"\n", sum);

        return 0;

}

int is_prime(int32_t s)
{
        int32_t res = !(s <= 2 || (s & 0x1) != 1);
        int32_t top = (int32_t) round(sqrt((float) s));

        for(int32_t i = 3; res && i <= top; i += ((res = !(s % i == 0)), 2))
                ;

        return (res || s == 2);
}

File: rust/summation_primes.rs




fn main () {

    let mut sum : i64 = 2;
    let max = 2000000;

    for i in 3..max {

        if isprime(i) {
            sum += i as i64;
        }
    }

    println!("{}", sum);

}

fn isprime(n : i32) -> bool {

    if (n <= 2) || (n & 0x1 == 0) {
        return n == 2;
    }

    let lim = (n as f32).sqrt().round() as i32;
    let mut i = 3;

    while i <= lim {

        if n % i == 0 {
            return false;
        }

        i += 2;
    }

    true
}

